sdk/storage/Azure.Storage.DataMovement.Files.Shares/src/PathScanner.cs

```diff
@@ -6,55 +6,80 @@
 using System.Runtime.CompilerServices;
 using System.Threading;
 using System.Threading.Tasks;
-using Azure.Core;
 using Azure.Storage.Common;
 using Azure.Storage.Files.Shares;
 using Azure.Storage.Files.Shares.Models;
+using Azure.Storage.Files.Shares.Specialized;
 
 namespace Azure.Storage.DataMovement.Files.Shares
 {
     internal class PathScanner
     {
         public static Lazy<PathScanner> Singleton { get; } = new Lazy<PathScanner>(() => new PathScanner());
 
-        public virtual async IAsyncEnumerable<ShareFileClient> ScanFilesAsync(
-            ShareDirectoryClient directory,
+        public virtual async IAsyncEnumerable<StorageResource> ScanAsync(
+            ShareDirectoryClient sourceDirectory,
+            ShareClient destinationShare,
+            ShareFileStorageResourceOptions sourceOptions,
+            ShareFileTraits traits,
             [EnumeratorCancellation] CancellationToken cancellationToken)
         {
-            await foreach ((ShareDirectoryClient _, ShareFileClient file) in
-                ScanAsync(directory, cancellationToken).ConfigureAwait(false))
-            {
-                if (file != default)
-                {
-                    yield return file;
-                }
-            }
-        }
-
-        public virtual async IAsyncEnumerable<(ShareDirectoryClient Dir, ShareFileClient File)> ScanAsync(
-            ShareDirectoryClient directory,
-            [EnumeratorCancellation] CancellationToken cancellationToken)
-        {
-            Argument.AssertNotNull(directory, nameof(directory));
+            Argument.AssertNotNull(sourceDirectory, nameof(sourceDirectory));
 
             Queue<ShareDirectoryClient> toScan = new();
-            toScan.Enqueue(directory);
+            toScan.Enqueue(sourceDirectory);
+
+            // Keep track of created Permission Keys to avoid creating duplicates.
+            ShareClient sourceShare = sourceDirectory.GetParentShareClient();
+            // Permissions keys <sourcePermissionKey, destinationPermissionKey>
+            Dictionary<string, string> permissionKeys = new();
 
             while (toScan.Count > 0)
             {
                 ShareDirectoryClient current = toScan.Dequeue();
                 await foreach (ShareFileItem item in current.GetFilesAndDirectoriesAsync(
+                    options: new() { Traits = traits },
                     cancellationToken: cancellationToken).ConfigureAwait(false))
                 {
+                    string destinationPermissionKey = string.Empty;
+                    if (destinationShare != default && item.PermissionKey != default)
+                    {
+                        // Check if the permission key is already created.
+                        // If not, create it on the Share.
+                        if (!permissionKeys.TryGetValue(item.PermissionKey, out string existingDestinationKey))
```

Comment:
Check options bag before copying permissions.

Reply:
Done. Preserved one level up in `ShareDirectoryStorageResourceContainer.GetStorageResourcesAsync`